src/engine/Minus.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@netpun.uni-freiburg.de)
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "./Operation.h"
+#include "./QueryExecutionTree.h"
+
+class Minus : public Operation {
+  enum class RowComparison {
+    EQUAL,
+    NOT_EQUAL_LEFT_SMALLER,
+    NOT_EQUAL_RIGHT_SMALLER
+  };
+
+ public:
+  Minus(QueryExecutionContext* qec, std::shared_ptr<QueryExecutionTree> left,
+        std::shared_ptr<QueryExecutionTree> right,
+        const std::vector<array<size_t, 2>>& matchedColumns);
+
+  virtual string asString(size_t indent = 0) const override;
+
+  virtual string getDescriptor() const override;
+
+  virtual size_t getResultWidth() const override;
+
+  virtual vector<size_t> resultSortedOn() const override;
+
+  ad_utility::HashMap<string, size_t> getVariableColumns() const;
+
+  virtual void setTextLimit(size_t limit) override {
+    _left->setTextLimit(limit);
+    _right->setTextLimit(limit);
+  }
+
+  virtual bool knownEmptyResult() override { return _left->knownEmptyResult(); }
+
+  virtual float getMultiplicity(size_t col) override;
+
+  virtual size_t getSizeEstimate() override;
+
+  virtual size_t getCostEstimate() override;
+
+  vector<QueryExecutionTree*> getChildren() override {
+    return {_left.get(), _right.get()};
+  }
+
+  /**
+   * @brief Joins a and b using the column defined int joinColumns, storing the
+   *        result in result. R should have width resultWidth (or be a vector
+   *        that should have resultWidth entries).
+   *        This method is made public here for unit testing purposes.
+   **/
+  template <int A_WIDTH, int B_WIDTH, int OUT_WIDTH>
+  static void computeMinus(const IdTable& a, const IdTable& b,
+                           const vector<array<size_t, 2>>& matchedColumns,
+                           IdTable* result);
+
+ private:
+  /**
+   * @brief Compares the two rows under the assumption that the first
+   * entries of the rows are equal.
+   */
+  template <int A_WIDTH, int B_WIDTH>
+  static RowComparison isRowEqSkipFirst(
+      const IdTableStatic<A_WIDTH>& a, const IdTableStatic<B_WIDTH>& b,
+      size_t ia, size_t ib, const vector<array<size_t, 2>>& matchedColumns);
+
+  std::shared_ptr<QueryExecutionTree> _left;
+  std::shared_ptr<QueryExecutionTree> _right;
+
+  vector<float> _multiplicities;
+  size_t _sizeEstimate;
+  bool _multiplicitiesComputed;
+
+  std::vector<array<size_t, 2>> _matchedColumns;
+
```

Comment:
It's interesting. It seems to be a good Idea in a codebase with lots of heavy data handling. I usually go for

* Inner Types
* Constructors
* Public Methods
* Private Methods
* Private Fields

Which focuses more on the ways a user can interact with the class (it's public methods). While still keeping its data storage consice at the bottom. I think that picking a fixed order is definitely a good idea. For me personally the order I'm used too works well, and is the least effort. If you'd rather put the fields at the top I'm fine with that as well, I don't think the actual position matters to much, as long as it's on one end of the file and clearly grouped.

Reply:
Your order also works, since one can easily find the ending of a class and work their way up.
I just realized that we have many types where it goes

```
private:
// private members

// private methods
``` 
And then the data is somewhere unfindably spread over the  middle.